### DISCUSSION TITLE
How to use ErrorOr...

### DISCUSSION THREAD

**Discussion by Jacrispys** [smallrye/smallrye-graphql#1449] [Q&A] [answered]
I am currently implementing a typesafe client for queries, when trying to add a ErrorOr handler to my class I get this error...
![image](https://user-images.githubusercontent.com/69219325/175135116-7d171eff-02b1-4aa0-9eb4-d8c6d783c04c.png)
It seems that the ErrorOr is messing up the syntax, but I'm not sure what type parameter I'm supposed to pass as the "location" to the ErrorOr object... Any help is greatly appreciated!

**Comment by phillip-kruger** [comment, marked as answer]
@jmartisk ^^^ ?

**Reply by t1** [reply]
It's difficult to understand your issue without seeing the code. Can you give us the relevant parts?

**Reply by Jacrispys** [reply]
> 

yeah for sure, basically I have a class
![image](https://user-images.githubusercontent.com/69219325/175369479-344103ef-9d3e-43c5-b26d-8766a68ebb45.png)
this returns data from a query and works fine, but when I add a field for `private ErrorOr<?> errors` it will cause the error shown above... I'm not sure what type I need to pass into the ErrorOr as well...
and here is the simple client API I use to wire it
![image](https://user-images.githubusercontent.com/69219325/175369829-931cedb8-fcfe-44f2-bc8f-c832e20a77aa.png)

**Reply by jmartisk** [reply]
You shouldn't add an `ErrorOr` field into the class, you should wrap the class into `ErrorOr` in your client API, so the query method becomes 
```
@Query("me")
ErrorOr<SelfUser> getSelfUser();
```

**Reply by t1** [reply]
It should also work for nested fields, as long as the surrounding object is present in the partial result.

**Reply by t1** [reply]
I wrote a minimalistic demo to show how the (test) GraphQL Client can read partial data via `ErrorOr` on a field: https://github.com/t1/error-or-demo

Maybe it's helpful. Otherwise, could you provide a reproducer, @Jacrispys ?

**Reply by Jacrispys** [reply]
> I wrote a minimalistic demo to show how the (test) GraphQL Client can read partial data via `ErrorOr` on a field: https://github.com/t1/error-or-demo
> 
> Maybe it's helpful. Otherwise, could you provide a reproducer, @Jacrispys ?

This is helpful, though i'm trying to throw an AuthorizationException when the query is called if the user doesn't use the correct headers... Any idea how I would throw an error from the query itself and not one of the fields? 


or maybe this is a better way of doing it and using and impl?

> You shouldn't add an `ErrorOr` field into the class, you should wrap the class into `ErrorOr` in your client API, so the query method becomes
> 
> ```
> @Query("me")
> ErrorOr<SelfUser> getSelfUser();
> ```

**Reply by Jacrispys** [reply]
> > I wrote a minimalistic demo to show how the (test) GraphQL Client can read partial data via `ErrorOr` on a field: https://github.com/t1/error-or-demo
> 
> > 
> 
> > Maybe it's helpful. Otherwise, could you provide a reproducer, @Jacrispys ?
> 
> 
> 
> This is helpful, though i'm trying to throw an AuthorizationException when the query is called if the user doesn't use the correct headers... Any idea how I would throw an error from the query itself and not one of the fields? 
> 
> 
> 
> 
> 
> or maybe this is a better way of doing it and using and impl?
> 
> 
> 
> > You shouldn't add an `ErrorOr` field into the class, you should wrap the class into `ErrorOr` in your client API, so the query method becomes
> 
> > 
> 
> > ```
> 
> > @Query("me")
> 
> > ErrorOr<SelfUser> getSelfUser();
> 
> > ```

For more clarification, I would like to have the ClientApi interface to be implemented into a impl of the interface where I can throw errors and execute other methods... it would also be nice to be able to get the graphql errors from the ClientApi and throw my own more specific errors.

**Reply by Jacrispys** [reply]
Fixed with this code...
![image](https://user-images.githubusercontent.com/69219325/175699317-720cc1c4-9820-4c1d-97d1-95bf0e38ef02.png)

**Reply by t1** [reply]
To me, this inheritance is confusing. I'd rename the `T4J` interface to `Api` and the `T4JImpl` class to `T4J` and not let it implement the interface. The `T4J` is the only one using the `Api`, so I'd make it a nested interface in `T4J`, so it's `T4J$Api`. Just my 2¢.
